File: tests/simulation_parameters/simulation_config_test.py
from simulation_parameters import SimulationConfig, TransientSimulationParameters, StepParameters


class TestSimulationConfig:

    def test_parse_analysis_and_step(self):
        # arrange
        directives = [
            ".TRAN 1u 1m",
            ".STEP R1 1k 10k 1k"
        ]
        # act
        config = SimulationConfig.from_xyce_directives(directives)
        # assert
        assert isinstance(config.analysis, TransientSimulationParameters)
        assert config.analysis.initial_step_value == "1u"
        assert config.step.enabled is True
        assert config.step.variable == "R1"

    def test_parse_only_analysis(self):
        # arrange
        directives = [".TRAN 1u 1m"]
        # act
        config = SimulationConfig.from_xyce_directives(directives)
        # assert
        assert isinstance(config.analysis, TransientSimulationParameters)
        assert config.step.enabled is False

    def test_parse_nothing(self):
        # arrange
        directives = ["* just a comment"]
        # act
        config = SimulationConfig.from_xyce_directives(directives)
        # assert
        assert config.analysis is None
        assert config.step.enabled is False

    def test_generate_directives(self):
        # arrange
        analysis = TransientSimulationParameters("1u", "1m", "", "", "", tuple())
        step = StepParameters(sweep_mode="LIN", variable="R1", start="1k", stop="10k", step="1k", enabled=True)
        config = SimulationConfig(analysis=analysis, step=step)
        # act
        directives = config.to_xyce_directives()
        # assert
        assert ".TRAN 1u 1m" in directives
        assert ".STEP R1 1k 10k 1k" in directives

File: tests/simulation_parameters/simulation_config.test.cpp
#include <algorithm>

#include <gtest/gtest.h>

#include "simulation_parameters/simulation_config.h"

TEST(SimulationConfigReplaceGroundChecks, from_xyce_directives_parses_disabled_statement) {
    // arrange / act
    const auto config = SimulationConfig::from_xyce_directives({".OP", ".PREPROCESS REPLACEGROUND FALSE"});
    // assert
    EXPECT_FALSE(config.replace_ground);
}

TEST(SimulationConfigReplaceGroundChecks, from_xyce_directives_defaults_to_true_without_statement) {
    // arrange / act
    const auto config = SimulationConfig::from_xyce_directives({".OP"});
    // assert
    EXPECT_TRUE(config.replace_ground);
}

TEST(SimulationConfigReplaceGroundChecks, to_xyce_directives_emits_disabled_statement) {
    // arrange
    const SimulationConfig config("OP", OpSimulationParameters(false, false, false, {}, "", "", false, "NODESET", "", {}, {}, std::nullopt), {}, {}, OptionParameters({}, {}, {}, {}, {}), {}, false);
    // act
    const auto directives = config.to_xyce_directives(NetlistTopology{});
    // assert
    const auto found = std::find(directives.begin(), directives.end(), ".PREPROCESS REPLACEGROUND FALSE");
    ASSERT_NE(found, directives.end());
}

TEST(SimulationConfigReplaceGroundChecks, disabled_state_round_trips_through_directives) {
    // arrange
    const SimulationConfig input("OP", OpSimulationParameters(false, false, false, {}, "", "", false, "NODESET", "", {}, {}, std::nullopt), {}, {}, OptionParameters({}, {}, {}, {}, {}), {}, false);
    // act
    const auto directives = input.to_xyce_directives(NetlistTopology{});
    const auto output = SimulationConfig::from_xyce_directives(directives);
    // assert
    EXPECT_FALSE(output.replace_ground);
}

TEST(SimulationConfigReplaceGroundChecks, enabled_state_round_trips_through_directives) {
    // arrange
    const SimulationConfig input("OP", OpSimulationParameters(false, false, false, {}, "", "", false, "NODESET", "", {}, {}, std::nullopt), {}, {}, OptionParameters({}, {}, {}, {}, {}), {}, true);
    // act
    const auto directives = input.to_xyce_directives(NetlistTopology{});
    const auto output = SimulationConfig::from_xyce_directives(directives);
    // assert
    EXPECT_TRUE(output.replace_ground);
}
